File: src/lib/datastructures/aho_corasick.ts
// Aho-Corasick solution
function AhoCorasick(words) {
	this.initWords(words);
}
AhoCorasick.prototype.Trie = function () {
	this.suffixLink = null;
	this.id = -1;
	this.next = new Map();
};
AhoCorasick.prototype.initWords = function (words) {
	this.root = new this.Trie();
	this.uniqueIds = 0;
	this.words = words;
	for (const word of words) {
		let p = this.root;
		for (const c of word) {
			if (!p.next.has(c)) {
				p.next.set(c, new this.Trie());
			}
			p = p.next.get(c);
		}
		if (p.id === -1) {
			p.id = this.uniqueIds++;
		}
	}
	this.buildAutomata();
};
AhoCorasick.prototype.buildAutomata = function () {
	let q = [];
	for (const [c, node] of this.root.next) {
		q.push(node);
		node.suffixLink = this.root;
	}
	while (q.length) {
		let nq = [];
		for (const curr of q) {
			for (const [c, node] of curr.next) {
				let ptr = curr.suffixLink;
				while (ptr !== this.root && !ptr.next.has(c)) {
					ptr = ptr.suffixLink;
				}
				node.suffixLink = ptr.next.get(c) ?? this.root;
				if (node.suffixLink.id !== -1) {
					node.id = node.suffixLink.id;
				}
				nq.push(node);
			}
		}
		q = nq;
	}
};
AhoCorasick.prototype.findMatches = function (target) {
	const matchMap = new Map();
	let ptr = this.root;
	for (let i = 0; i < target.length; ++i) {
		const c = target[i];
		while (ptr !== this.root && !ptr.next.has(c)) {
			ptr = ptr.suffixLink;
		}
		ptr = ptr.next.get(c) ?? this.root;
		if (ptr.id !== -1) {
			matchMap.set(i, this.words[ptr.id]);
		}
	}
	return matchMap;
};
export { AhoCorasick };

//alternative version of aho corasick
(function () {
	function AhoCorasick2(words) {
		this.root = new this.Trie();
		this.uniqueIds = 0;
		this.insert(words);
		this.buildAutomata();
	}
	AhoCorasick2.prototype.Trie = function () {
		this.suffixLink = null;
		this.id = -1;
		this.list = new Map();
	};
	AhoCorasick2.prototype.insert = function (words) {
		for (const word of words) {
			let p = this.root;
			for (const c of word) {
				if (!p.list.has(c)) {
					p.list.set(c, new this.Trie());
				}
				p = p.list.get(c);
			}
			if (p.id === -1) {
				p.id = this.uniqueIds++;
			}
		}
	};
	AhoCorasick2.prototype.buildAutomata = function () {
		let q = [];
		for (const [c, trie] of this.root.list) {
			q.push(trie);
			trie.suffixLink = this.root;
		}
		while (q.length) {
			let nq = [];
			for (const curr of q) {
				for (const [c, trie] of curr.list) {
					let ptr = curr.suffixLink;
					while (ptr !== this.root && !ptr.list.has(c)) {
						ptr = ptr.suffixLink;
					}
					trie.suffixLink = ptr.list.get(c) ?? this.root;
					if (trie.suffixLink.id !== -1) {
						trie.id = trie.suffixLink.id;
					}
					nq.push(trie);
				}
			}
			q = nq;
		}
	};
	AhoCorasick2.prototype.findMatches = function (target) {
		const matchMap = new Map();
		let ptr = this.root;
		for (let i = 0; i < target.length; ++i) {
			const c = target[i];
			while (ptr !== this.root && !ptr.list.has(c)) {
				ptr = ptr.suffixLink;
			}
			ptr = ptr.list.get(c) ?? this.root;
			if (ptr.id !== -1) {
				matchMap.set(i, ptr.id);
			}
		}
		return matchMap;
	};
})();

File: src/lib/datastructures/aho_corasick.py
from typing import List
from queue import Queue

print(f"Identifier->{__name__}")


def main():
    # https://www.youtube.com/watch?v=g_wlZ9IhbTs
    # src on why this should be done
    print("This is an importable data structure. The aho-corasick automaton.")


if __name__ == "__main__":
    main()


class AhoCorasick:
    class InternalTrie:
        def __init__(self):
            self.suffixLink = None
            self.id = -1
            self.next = {}

    def __init__(self, words: List[str] = []):
        self.initWords(words)

    def initWords(self, words):
        self.root = self.InternalTrie()
        self.words = words
        self.uniqueIds = 0

        for word in words:
            p = self.root

            for c in word:
                if c not in p.next:
                    p.next[c] = self.InternalTrie()
                p = p.next.get(c)

            if p.id == -1:
                p.id = self.uniqueIds
                self.uniqueIds += 1

        self.buildAutomata()

    def buildAutomata(self):
        q = Queue()
        for c, node in self.root.next.items():
            q.put(node)
            node.suffixLink = self.root

        while not q.empty():
            curr = q.get(0)
            for c, node in curr.next.items():
                ptr = curr.suffixLink
                while ptr is not None and c not in ptr.next:
                    ptr = ptr.suffixLink

                node.suffixLink = ptr.next.get(c) if ptr is not None and c in ptr.next else self.root
                if node.suffixLink.id != -1:
                    node.id = node.suffixLink.id

                q.put(node)

    def findMatches(self, target: str) -> dict:
        matchMap = {}
        ptr = self.root
        for i in range(len(target)):
            c = target[i]

            while ptr is not None and c not in ptr.next:
                ptr = ptr.suffixLink
            ptr = ptr.next.get(c) if ptr is not None and c in ptr.next else self.root

            if ptr.id != -1:
                matchMap[i] = self.words[ptr.id]

        return matchMap

    def printRoot(self) -> None:
        print("printRoot:: start **********")
        q = Queue()
        q.put([self.root, ''])
        words = []
        while not q.empty():
            curr, currStr = q.get(0)
            if curr.id != -1:
                words.append(currStr)
            for c, node in curr.next.items():
                print(f"c: {c} nodeId: {node.id}")
                q.put([node, currStr + c])

        print(words)
        print("printRoot:: end ********** ")
